File: frontend/nurse-app/src/app/services/patient.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Patient } from '../models/patient';
import { Observable } from 'rxjs/internal/Observable';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class PatientService {

  constructor(private httpClient: HttpClient) { }

  getPatients(): Observable<Patient[]> {
    return this.httpClient.get<Patient[]>(environment.url + '/getPatients', { withCredentials: true });
  }

  getPatientById(patient_id: string) : Observable<Patient> {
    return this.httpClient.post<Patient>(environment.url + "/getPatientById", patient_id);
  }

  addPatient(patient: Patient) {
    return this.httpClient.post(environment.url + "/addPatient", patient);
  }

  addPatientCondition(patient_id:string, condition: string) {
    return this.httpClient.post(environment.url + "/patient/conditions", { patient_id: patient_id, condition: condition})
  }

  getPatientConditions(patient_id: string) {
    return this.httpClient.get(environment.url + "/patient/" + patient_id + "/conditions");
  }

  deletePatientCondition(patient_id: string, condition_name: string) {
    return this.httpClient.delete(environment.url + "/patient/" + patient_id + "/conditions/delete",
    { params: {condition_name: condition_name}})
  }

  searchPatientsByCondition(conditionName: string): Observable<any> {
    const url = `${environment.url}/patient/conditions/search?condition_name=${conditionName}`;
    return this.httpClient.get(url);
  }

  searchPatient(searchTerm: string): Observable<Patient[]> {
    return this.httpClient.post<Patient[]>(environment.url + "/searchPatient", {search_term: searchTerm})
  }

  updatePatient(patient: Patient) {
    return this.httpClient.post(environment.url + "/updatePatient", patient);
  }

  deletePatient(patient_id: string) {
    return this.httpClient.post(environment.url + "/deletePatient", patient_id);
  }

  searchPatientByNotes(searchTerm: string) {
    return this.httpClient.post<Patient[]>(environment.url + "/searchNotes", {search_term: searchTerm})
  }
}

File: backend/services/patient_service.py
from models.patient import Patient
from repositories import patient_repository
from geopy.geocoders import Nominatim

def get_patients():
    return patient_repository.get_patients()


def get_patient_by_id(patient_id):
    return patient_repository.get_patient_by_id(patient_id)


def add_patient(patient):
    new_patient = Patient()
    new_patient.first_name = patient["first_name"]
    new_patient.last_name = patient["last_name"]
    new_patient.address = patient["address"]
    geolocator = Nominatim(user_agent="my_request")
    new_patient.location = geolocator.geocode(new_patient.address)
    return patient_repository.add_patient(new_patient)


def update_patient(patient):
    patient_updated = Patient()
    patient_updated.patient_id = patient["id"]
    patient_updated.first_name = patient["first_name"]
    patient_updated.last_name = patient["last_name"]
    patient_updated.address = patient["address"]
    patient_updated.notes = patient["notes"].upper()
    patient_updated.phone_number = patient["phone_number"]
    geolocator = Nominatim(user_agent="my_request")
    patient_updated.location = geolocator.geocode(patient_updated.address)
    return patient_repository.update_patient(patient_updated)


def add_patient_condition(patient_id, condition):
    return patient_repository.add_patient_condition(patient_id, condition)


def get_patient_conditions(patient_id):
    return patient_repository.get_patient_conditions(patient_id)


def delete_patient_condition(patient_id, condition):
    return patient_repository.delete_patient_condition(patient_id, condition)


def search_patients_by_condition(search_term):
    return patient_repository.search_patients_by_condition(search_term)


def delete_patient(patient_id):
    return patient_repository.delete_patient(patient_id)


def search_patient(search_term):
    return patient_repository.search_patient(search_term)


def search_patients_by_notes(search_term):
    return patient_repository.search_patients_by_notes(search_term)